test/sparse/higherorderfns.jl

```diff
@@ -0,0 +1,254 @@
+            @test broadcast!(+, Q, X, Y, Z) == sparse(broadcast!(+, fQ, fX, fY, fZ))
+            # --> test broadcast! entry point / *-like zero-preserving op
+            fQ = broadcast(*, fX, fY, fZ); Q = sparse(fQ)
+            broadcast!(*, Q, X, Y, Z); Q = sparse(fQ) # warmup for @allocated
+            @test (@allocated broadcast!(*, Q, X, Y, Z)) == 0
+            @test broadcast!(*, Q, X, Y, Z) == sparse(broadcast!(*, fQ, fX, fY, fZ))
+            # --> test broadcast! entry point / not zero-preserving op
+            fQ = broadcast(f, fX, fY, fZ); Q = sparse(fQ)
+            broadcast!(f, Q, X, Y, Z); Q = sparse(fQ) # warmup for @allocated
+            @test_broken (@allocated broadcast!(f, Z, X, Y)) == 0
+            # the preceding test allocates 16 bytes in the entry point for broadcast!, but
+            # none of the earlier tests of the same code path allocate. no allocation shows
+            # up with --track-allocation=user. allocation shows up on the first line of the
+            # entry point for broadcast! with --track-allocation=all, but that first line
+            # almost certainly should not allocate. so not certain what's going on.
+            @test broadcast!(f, Q, X, Y, Z) == sparse(broadcast!(f, fQ, fX, fY, fZ))
+            # --> test shape checks for both braodcast and broadcast! entry points
+            try Base.Broadcast.check_broadcast_indices(indices(Q), spzeros((shapeX .- 1)...), Y, Z)
+            catch @test_throws DimensionMismatch broadcast!(f, Q, spzeros((shapeX .- 1)...), Y, Z) end
+        end
+    end
+end
+
+# Older tests of sparse broadcast, now largely covered by the tests above
+@testset "assorted tests of sparse broadcast over two input arguments" begin
+    N, p = 10, 0.3
+    A, B, CF = sprand(N, N, p), sprand(N, N, p), rand(N, N)
+    AF, BF, C = Array(A), Array(B), sparse(CF)
+
+    @test A .* B == AF .* BF
+    @test A[1,:] .* B == AF[1,:] .* BF
+    @test A[:,1] .* B == AF[:,1] .* BF
+    @test A .* B[1,:] == AF .*  BF[1,:]
+    @test A .* B[:,1] == AF .*  BF[:,1]
+
+    @test A .* B == AF .* BF
+    @test A[1,:] .* BF == AF[1,:] .* BF
+    @test A[:,1] .* BF == AF[:,1] .* BF
+    @test A .* BF[1,:] == AF .*  BF[1,:]
+    @test A .* BF[:,1] == AF .*  BF[:,1]
+
+    @test A .* B == AF .* BF
+    @test AF[1,:] .* B == AF[1,:] .* BF
+    @test AF[:,1] .* B == AF[:,1] .* BF
+    @test AF .* B[1,:] == AF .*  BF[1,:]
+    @test AF .* B[:,1] == AF .*  BF[:,1]
+
+    @test A .* B == AF .* BF
+    @test A[1,:] .* B == AF[1,:] .* BF
+    @test A[:,1] .* B == AF[:,1] .* BF
+    @test A .* B[1,:] == AF .*  BF[1,:]
+    @test A .* B[:,1] == AF .*  BF[:,1]
+
+    @test A .* 3 == AF .* 3
+    @test 3 .* A == 3 .* AF
+    #@test A[1,:] .* 3 == AF[1,:] .* 3
+    @test all(A[1,:] .* 3 .== AF[1,:] .* 3)
+    #@test A[:,1] .* 3 == AF[:,1] .* 3
+    @test all(A[:,1] .* 3 .== AF[:,1] .* 3)
+    #TODO: simple comparation with == returns false because the left side is a (two-dimensional) SparseMatrixCSC
```

Comment:
(Is this a request to clear this TODO in this pull request? Or a request to clear it later? Thanks!)